File: Phase-3/frontend/types/conversation.ts
/**
 * Conversation and message types
 */

export interface Message {
  id: number
  conversation_id: number
  role: "user" | "assistant" | "system"
  content: string
  created_at: string
}

export interface ToolExecution {
  id: number
  tool_name: string
  parameters: Record<string, unknown>
  result?: Record<string, unknown> | null
  status: "pending" | "success" | "failed"
  error_message?: string | null
  created_at: string
}

export interface Conversation {
  id: number
  user_id: string
  title?: string | null
  is_active: boolean
  created_at: string
  updated_at: string
  messages?: Message[]
}

export interface ConversationCreate {
  title?: string
  initial_message?: string
}

export interface SendMessageRequest {
  content: string
}

export interface SendMessageResponse {
  user_message: Message
  assistant_message: Message
  tool_executions: ToolExecution[]
}

File: Phase-5/backend/src/schemas/conversation.py
"""Conversation Pydantic schemas for API."""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class ConversationCreate(BaseModel):
    """Schema for creating a new conversation."""

    title: Optional[str] = Field(None, max_length=255)
    initial_message: Optional[str] = Field(None, description="Optional first message to send")


class MessageCreate(BaseModel):
    """Schema for sending a message in a conversation."""

    content: str = Field(..., min_length=1, description="Message content from user")


class MessageResponse(BaseModel):
    """Schema for message response."""

    id: int
    conversation_id: int
    role: str
    content: str
    created_at: datetime

    class Config:
        """Pydantic config."""
        from_attributes = True


class ToolExecutionResponse(BaseModel):
    """Schema for tool execution response."""

    id: int
    tool_name: str
    parameters: dict
    result: Optional[dict] = None
    status: str
    error_message: Optional[str] = None
    created_at: datetime

    class Config:
        """Pydantic config."""
        from_attributes = True


class SendMessageResponse(BaseModel):
    """Schema for send message endpoint response."""

    user_message: MessageResponse
    assistant_message: MessageResponse
    tool_executions: list[ToolExecutionResponse]


class ConversationResponse(BaseModel):
    """Schema for conversation response."""

    id: int
    user_id: str
    title: Optional[str]
    is_active: bool
    created_at: datetime
    updated_at: datetime
    messages: Optional[list[MessageResponse]] = None

    class Config:
        """Pydantic config."""
        from_attributes = True
